src/components/Form/Widgets/FileWidget.tsx

```diff
@@ -1,248 +1,429 @@
-				const uploadErrors = fileRejections[0].errors.map(
-					({ message }) => `${fileRejections[0].file.name}: ${message}`,
-				);
-				setErrors(uploadErrors);
-				return onChange({ uploadErrors });
-			}
-			const dataUrls: string[] = [];
-			acceptedFiles.forEach((file, i) => {
+		(acceptedFiles: AcceptedFile[], rejectedFiles) => {
+			acceptedFiles.forEach((file) => {
 				const reader = new FileReader();
 				reader.onerror = () => {
-					const uploadErrors = [`Failed to upload ${file.name}`];
-					setErrors(uploadErrors);
-					onChange({ uploadErrors });
+					if (file.errors == null) {
+						file.errors = [];
+					}
+					file.errors.push({ code: undefined, message: 'Failed to upload' });
+				};
+				reader.onloadstart = () => {
+					file.loadingPercentage = 0;
 				};
-				reader.onloadstart = () => setLoadingPercentage(0);
 				reader.onprogress = (event) => {
-					setLoadingPercentage(Math.floor((100 * event.loaded) / event.total));
+					file.loadingPercentage = Math.floor(
+						(100 * event.loaded) / event.total,
+					);
+					if (file.loadingPercentage === 100) {
+						delete file.loadingPercentage;
+					}
 				};
 				reader.onload = () => {
 					if (typeof reader.result !== 'string' || !reader.result || !file) {
 						return;
 					}
 					const base64Data = reader.result.split(',')[1];
+					file.base64Data = base64Data;
 					const dataUrl = `data:${file.type};name=${file.name};base64,${base64Data}`;
-					if (i === acceptedFiles.length - 1) {
-						setLoadingPercentage(undefined);
-					}
-					if (!multiple) {
-						onChange({
-							dataUrl,
-							uploadedFile: file,
-						});
+					if (multiple) {
+						// NOTE: JSONSchema array data-url does not expect objects but only strings[]
+						// see: https://github.com/rjsf-team/react-jsonschema-form/blob/297dac059fdf64fd1453bebb8366f0602c722f90/packages/utils/src/schema/isFilesArray.ts#L24
+						onChange(
+							[...files, ...acceptedFiles]
+								.filter(
+									(file) =>
+										file.loadingPercentage === null && !file.errors?.length,
+								)
+								.map(
+									(file) =>
+										`data:${file.type};name=${file.name};base64,${base64Data}`,
+								),
+						);
 						return;
 					}
-					// NOTE: JSONSchema array data-url does not expect objects but only strings[]
-					// see: https://github.com/rjsf-team/react-jsonschema-form/blob/297dac059fdf64fd1453bebb8366f0602c722f90/packages/utils/src/schema/isFilesArray.ts#L24
-					dataUrls.push(dataUrl);
-					onChange(dataUrls);
+					onChange({
+						dataUrl,
+						uploadedFile: file,
+					});
 				};
 				reader.readAsDataURL(file);
 			});
+			if (multiple) {
+				setFiles([
+					...files,
+					...acceptedFiles,
+					...rejectedFiles.map(
+						(rejectedFile) =>
+							({
+								...rejectedFile.file,
+								// for some reason, spreading rejectedFile.file does not spread these properties
+								name: rejectedFile.file.name,
```

Comment:
It's expected as the interfaces are different.

```
type AcceptedFile = File & {
    loadingPercentage?: number;
    base64Data?: string;
    errors?: Array<{ code: ErrorCode | undefined; message: string }>;
};

interface FileRejection {
    file: File;
    errors: FileError[];
}
```

The issue isn't with the spread operator failing to spread properties;
rather, these properties simply don't exist on the object.

Additionally, why are we attempting to display the file if it was rejected?
If the file is rejected, I would expect the widget not load the file but instead only display the error.

Reply:
The idea is to show the user a list of the files they tried to upload and the reasons the uploads failed per file. I will send the design in zulip, seeing it should be helpful